client/my-sites/site-settings/site-icon-setting/index.jsx

```diff
@@ -40,11 +51,45 @@ class SiteIconSetting extends Component {
 
 	showModal = () => this.toggleModal( true );
 
-	setSiteIcon = ( media ) => {
-		// [TODO]: Handle setting site icon
-		console.log( media ); // eslint-disable-line no-console
+	editSelectedMedia = ( media ) => {
+		if ( media ) {
+			this.props.onEditSelectedMedia();
+		}
+	};
 
+	setSiteIcon = ( error, blob ) => {
 		this.hideModal();
+		this.props.resetAllImageEditorState();
+
+		if ( error || ! blob ) {
+			return;
+		}
+
+		const { siteId } = this.props;
+		const selectedItem = head( MediaLibrarySelectedStore.getAll( siteId ) );
+		if ( ! selectedItem ) {
+			return;
+		}
+
+		const transientMediaId = uniqueId( 'site-icon' );
+
+		MediaActions.add( siteId, {
+			ID: transientMediaId,
+			fileName: `cropped-${ selectedItem.file }`,
+			fileContents: blob
+		} );
+
+		const checkUploadComplete = () => {
+			const transientMedia = MediaStore.get( siteId, transientMediaId );
+			if ( isItemBeingUploaded( transientMedia ) ) {
+				return;
+			}
+
+			MediaStore.off( 'change', checkUploadComplete );
+			alert( transientMedia.ID );
+		};
+
+		MediaStore.on( 'change', checkUploadComplete );
```

Comment:
>I'm wondering if we could have a "transition" state, where we would move the lines from 74:92 here to a redux thunk action creator. `MediaActions` will be used like an API abstraction.

I'm hesitant because we don't really know what we want the final solution to look like. We couldn't (or rather, _shouldn't_) swap out the Flux subscription with a subscription to the Redux store in the action creator. A Redux middleware could work, but given how infrequently this action occurs, it seems overkill (and has a performance impact). Rather, if I was to imagine a future solution, I think we'd subscribe to the Redux store just as such here (via `this.context.store.subscribe`), calling equivalent selectors to `get` and `isItemBeingUploaded`.

There was some discussion about this at p1480623194001494-slack-calypso-framework , wherein it was concluded that this approach will technically lead to a "zombie" view ([akin to a common Backbone.js problem](https://lostechies.com/derickbailey/2011/09/15/zombies-run-managing-page-transitions-in-backbone-apps/)) though an intentional one that we're confident will eventually unsubscribe at the point in time we'd want it to (i.e. after the request finishes).